src/fastaq_handler.cpp

```diff
@@ -1,176 +1,77 @@
-        std::exit(EXIT_FAILURE);
-    }
-    try {
-        if (filepath.substr(filepath.length() - 2) == "gz") {
-            inbuf.push(boost::iostreams::gzip_decompressor());
-            gzipped = true;
-        }
-        inbuf.push(fastaq_file);
-    } catch (const boost::iostreams::gzip_error& e) {
-        std::cerr << "Problem transfering file contents to boost stream: " << e.what()
-                  << '\n';
+    this->fastaq_file = gzopen(filepath.c_str(), "r");
+    if (this->fastaq_file == nullptr) {
+        throw "Unable to open " + this->filepath;
     }
+    this->inbuf = kseq_init(this->fastaq_file);
 }
 
-FastaqHandler::~FastaqHandler() { close(); }
+FastaqHandler::~FastaqHandler() { this->close(); }
 
-bool FastaqHandler::eof()
-{
-    int c = instream.peek();
-    return (c == EOF);
-}
+bool FastaqHandler::eof() const { return ks_eof(this->inbuf->f); }
 
 void FastaqHandler::get_next()
 {
-    // cout << "next ";
-    if (!line.empty() and (line[0] == '>' or line[0] == '@')) {
-        // cout << "read name line " << num_reads_parsed << " " << line << endl;
-        name = line.substr(1);
-        ++num_reads_parsed;
-        read.clear();
-    }
-
-    while (getline(instream, line).good()) {
-        if (!line.empty() and line[0] == '+') {
-            // skip this line and the qual score line
-            getline(instream, line);
-            // cout << "qual score line ." << line << "." << endl;
-        } else if (line.empty() || line[0] == '>' || line[0] == '@') {
-            if (!read.empty()
-                or line.empty()) // ok we'll allow reads with no name, removed
-            {
-                return;
-            }
-            // cout << num_reads_parsed << " " << line << endl;
-            name = line.substr(1);
-            ++num_reads_parsed;
-            read.clear();
-        } else {
-            // cout << "read line ." << line << "." << endl;
-            read += line;
-        }
-    }
-}
-
-void FastaqHandler::skip_next()
-{
-    // cout << "skip ";
-    if (!line.empty() and (line[0] == '>' or line[0] == '@')) {
-        ++num_reads_parsed;
-    }
-
-    while (getline(instream, line).good()) {
-        if (!line.empty() and line[0] == '+') {
-            // skip this line and the qual score line
-            getline(instream, line);
-            // cout << "qual score line ." << line << "." << endl;
-        } else if (line.empty() or line[0] == '>' or line[0] == '@') {
-            return;
-        }
-    }
-}
-
-void print(std::ifstream& infile)
-{
-    char file;
-    std::vector<char> read;
-    uint i = 0;
-
-    // Read infile to vector
-    while (!infile.eof()) {
-        infile >> file;
-        read.push_back(file);
-    }
+    int read_status = kseq_read(this->inbuf);
 
-    // Print read vector
-    for (i = 0; i < read.size(); i++) {
-        std::cout << read[i];
+    bool no_more_reads_available = read_status == -1;
+    if (no_more_reads_available) {
+        return;
```

Comment:
What do you think about throwing a `std::out_of_bounds` exception if the user try to read sequences when there are no more sequences? Here we don't do anything, i.e. `this->name` and `this->read` refer to the last sequence read, it seems to me a `std::out_of_bounds` would be better? I am not sure, what do you think?

Reply:
Hmm, this is a good point. Probably a good idea to throw an exception. It scares me to think we could be thinking we are getting more reads when we are actually just reusing the last read multiple times (I think we have tests that cover this, but throwing an exception is probably best).

This raises another thing that has been annoying me. I feels like `get_next` would be best to return some `Record` class for the read, rather than invisibly altering internal member variables. However, this is a much larger change...

Reply:
> It scares me to think we could be thinking we are getting more reads when we are actually just reusing the last read multiple times (I think we have tests that cover this, but throwing an exception is probably best).

Yeah, it also now scares me... the "convention" of `get_next()` functions is to throw errors or return a special value when there is nothing else to get, and we might get into trouble in the future if we don't remember that this `get_next()` don't do this...

> This raises another thing that has been annoying me. I feels like get_next would be best to return some Record class for the read, rather than invisibly altering internal member variables. However, this is a much larger change...

Yeah, I also agree... Maybe not a Record class, that does not exist in `pandora`, so we would need to create and refactor everywhere where `get_next()` is called, but maybe just a tuple with the information about what was read, or a `Seq` object (https://github.com/rmcolq/pandora/blob/master/include/seq.h).

PS: returning a `Seq` object won't work because the FastaqHandler handles more than fasta or fastq files. This might be sth important we might have missed, but it seems things did not break, so all is actually fine.
What I mean is that FastqHandler objects are used to handle PRG files: https://github.com/rmcolq/pandora/blob/6f6af4b7a137a158a97806e64c58ad8febb9ccc3/src/utils.cpp#L116-L138
... and PRG files are fasta-like files, but PRG sequences != DNA. I think some libraries might break if we try to read PRG files as fasta files (because of the numbers, spaces, etc...). Good thing that `kseq` does not seem to break, and handle PRG sequences just fine... I think SeqAn would break.
We definitely can't return a `Seq` object because when a `Seq` is created, `pandora` wants to automatically sketch it, which does not work for PRG sequences...

Bottom line is that I also agree with you on this point, and could be improved, but I am not sure if we should change anything now...

Reply:
PS: I agree the Record class to concretely represent a Record in a Fasta/q file is the best solution, but also the one needing most work... was trying to find some middle-ground solutions... I guess whatever you decide to do is fine by me!